[llvm-objcopy] Support SREC output format

quic-areg opened the pull request:
Adds a new output target "srec" to write SREC files from ELF inputs.

https://en.wikipedia.org/wiki/SREC_(file_format)

quic-areg committed:
[llvm-objcopy] Support SREC output format

Adds a new output target "srec" to write SREC files from ELF inputs.

https://en.wikipedia.org/wiki/SREC_(file_format)

quic-areg commented:
@MaskRay @jh7370 @JakeEhrlich

jh7370 reviewed (commented):
Thank you for your contribution! This is a very large patch, and I won't have time to review it until the new year. In the meantime, some high-level comments:
1) Please make sure to update the llvm-objcopy documentation. You can find it at llvm/docs/CommandGuide/llvm-objcopy.rst.
2) I assume you have a need for this functionality in your company/team/...? We don't generally accept patches for new functionality if there is no concrete use case.
3) It may be better to split this into a sequence of PRs, which incrementally add the functionality, rather than trying to add it all at once. For example, is it possible to support an empty ELF file with no sections/symbols/relocations/program headers in a first patch, and then add support for different section types etc in follow-on PRs?

quic-areg commented:
Thank you for your feedback. 

This functionality was mentioned in[ EuroLLVM 2023](https://discourse.llvm.org/t/llvm-embedded-toolchains-eurollvm-2023-round-table-summary/70700): 

> Support for Motorola srec format (GNU objcopy -O srec) would be useful for some projects wanting to transition from GNU.

SREC is used in our team as an intermediate format while generating memory images. Adding support would help them transition to LLVM.

I have updated the documentation to reflect the new output format.

I also simplified this patch like you suggested. It will only write the header record and ignore any sections.

MaskRay commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +390,57 @@ class IHexWriter : public Writer {
+  };
+};
+
+/// FIXME: This is currently a contrived writer that only writes the
```
If we decide to accept srec support, I think the MVP needs to support writing data.

quic-areg replied:
Done

MaskRay commented on llvm/test/tools/llvm-objcopy/ELF/Inputs/srec-elf-sections.yaml:
```diff
@@ -0,0 +1,48 @@
+# FIXME: Currently the sections in this file are ignored and we
```
Tests from the original ihex patch (https://reviews.llvm.org/D60270) may be useful.

quic-areg replied:
The tests in this patch are based on the ones in the ihex patch

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -0,0 +1,11 @@
+# RUN: yaml2obj %p/Inputs/srec-elf-sections.yaml -o %t
```
The test now uses multiple files

jh7370 replied:
The input file is only used in one single test. You can have multiple different YAML inputs within the same test file using the --docnum option. See many other tests that use this option for examples of this.

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -0,0 +1,11 @@
+# RUN: yaml2obj %p/Inputs/srec-elf-sections.yaml -o %t
+# RUN: llvm-objcopy -O srec %t - | FileCheck %s
+
+# The record type for the header should be S0 with a 2 byte address
```
Done

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -0,0 +1,11 @@
+CHECK: S00400002DCE
+CHECK-EMPTY:
+
+# FIXME: write the rest of the records
```
Done

statham-arm commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -324,6 +324,14 @@ Expected<IHexRecord> IHexRecord::parse(StringRef Line) {
   return Rec;
 }
 
+static uint64_t sectionLMA(const SectionBase *Sec) {
```
I'm curious about why this helper function is needed for SREC, if it didn't already exist for IHEX. The two seem similar to me: surely for both formats you want to calculate the memory addresses of segments the same way.

Does this patch agree with the way IHEX does it? If not, which do you think is right?

statham-arm replied:
In fact, now I look closer, the way IHEX does it is in the `sectionPhysicalAddr` function just below this new one. What's the reasoning for needing to do it differently?

quic-areg replied:
I see different results from gnu objcopy when using sectionPhysicalAddr. 

example using `ihex-elf-segments.yaml`:

GNU:
```
S00400007883
S214000000000102030405060708090A0B0C0D0E0F73
S20900001010111213148C
S20F00FFF83031323334353637383940AC
S20801010040414243EF
S20F10FFF850515253545556575859603C
S20720FFF8000000E1
S804100000EB

address data 

000000 000102030405060708090A0B0C0D0E0F
000010 1011121314
00FFF8 3031323334353637383940
010100 40414243
10FFF8 5051525354555657585960
20FFF8 000000
```

LLVM(sectionPhysicalAddr):
```
S00400007883
S214100000000102030405060708090A0B0C0D0E0F63
S20910001010111213147C
S20F10001830313233343536373839407B
S20810002840414243B9
S20F100030505152535455565758596003
S20720FFF8000000E1
S804100000EB

address data 

100000 000102030405060708090A0B0C0D0E0F
100010 1011121314
100018 3031323334353637383940
100028 40414243
100030 5051525354555657585960
20FFF8 000000
```

looking at the LMA values in objdump -h output, I think that the former is correct
```
objdump -h 1.o 

1.o:     file format elf64-x86-64

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .text         00000015  0000000000000000  0000000000000000  00000078  2**3
  1 .data1        0000000b  000000000000fff8  000000000000fff8  00000090  2**3
  2 .data2        00000004  0000000000010100  0000000000010100  000000a0  2**3
  3 .data3        0000000b  000000000010fff8  000000000010fff8  000000a8  2**3
```

statham-arm replied:
OK, so in this example ELF file, there are _three_ address layouts. (`llvm-readelf` or GNU `readelf` are better for showing this than `objdump`.)

Going by the addresses in the section header table, this image is highly non-contiguous, with `.text` going at 0, some data around 0x10000, some more data just below 0x200000 and more still just below 0x300000. Each section is only a few bytes long, and the gaps between them are many Kb or even Mb.

But then those sections are packed into a single `PT_LOAD` segment much more tightly, so that the only gaps between them are a few bytes for alignment. That segment has two different addresses: a virtual address (`p_vaddr`) of 0xF00000000, and a physical address (`p_paddr`) of 0x100000.

GNU objcopy, in both `ihex` and `srec` output mode, is using the layout from the section addresses, so that it emits records with addresses near 0, near 0x10000, near 0x200000 and near 0x300000.

`llvm-objcopy -O ihex` is using the layout from the segment's physical address, so that everything it emits is tightly packed near 0x100000.

`llvm-objcopy -O srec` with your patch is using the section addresses again, so that LLVM's two hex formats produce different addresses.

(And nothing is using the _virtual_ address of the segment, which is just as well, because it's greater than 2^32, and wouldn't fit into these 32-bit hex file formats in any case.)

I can see that you might want to be able to generate hex output based on the section addresses. But I think that the choice of addressing scheme should not depend on which hex format you're outputting. The two hex formats are used for similar purposes; there's no reason why one should be right for `srec` and the other one for `ihex`.

So I'd prefer that we use the same section-address function for both hex formats, and separately, introduce an option to control how that function decides on addresses.

quic-areg replied:
Thank you for the explanation.

Updated this patch to use the same section address function as ihex and updated a test to reflect this.

statham-arm reviewed (changes requested):
This looks _almost_ fine to me now. I have only one remaining nitpick, and I apologise for not having spotted it the first time.

quic-areg commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2813,6 +2821,235 @@ Error IHexWriter::finalize() {
+}
+
+Error SRECWriter::checkSection(const SectionBase &S) const {
+  if (!isUInt<32>(S.Addr + S.Size - 1))
```
I agree it is a good idea. I added sign-extended addresses and updated the tests.

I also fixed a bug I found when updating the tests:

```
## Terminator should contain the entry point
# RUN: llvm-objcopy -O srec --set-start=0xF0000000 %t \
#   RUN: --only-section=.dummy - 2>&1 | FileCheck %s --check-prefix=ENTRY
```

Previously this would cause a crash because the entry point was not being considered in determining the record type and did not fit in an S1/S9 record. GNU objcopy has a similar bug, but it truncates the start address to fit 16 bits instead of choosing an appropriate record type.

statham-arm reviewed (approved):
LGTM this time!

jh7370 commented:
I'd like to take a look at this before this gets merged, please, but I'm out of time to review this today. I should be able to review this tomorrow.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2813,6 +2813,244 @@ Error IHexWriter::finalize() {
+  return Error::success();
+}
+
+Error SRECWriter::finalize() {
```
This function is nearly identical to the `IHexWriter::finalize` function. Please refactor the two to share the code as much as possible.

A basic programming principle is don't repeat yourself: avoid duplicating code as much as possible, by using functions/classes etc to factor out the common code into a single place.

quic-areg replied:
I added a new base class for srec and ihex called `ASCIIHexWriter` to factor out common stuff.  

Since they share the same code, ihex errors now start with lower case.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  ArrayRef<uint8_t> Data;
+  SRecLineData toString() const;
+  uint8_t getCount() const;
+  // get address size in characters
```
Comment style, here and throughout this file.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  };
+};
+
+/// The real writer
```
I don't think this comment adds anything. Please delete it.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  Error checkSection(const SectionBase &S) const;
+
+public:
+  ~SRECWriter() = default;
```
Is this necessary?

quic-areg replied:
No. Removed.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  ~SRECWriter() = default;
+  Error finalize() override;
+  Error write() override;
+  SRECWriter(Object &Obj, raw_ostream &OS, StringRef OutputFile)
```
It's more common to write constructors first.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  size_t writeTerminator(uint8_t *Buf, uint8_t Type);
+  Error checkSection(const SectionBase &S) const;
+
+public:
```
It's more common to put public stuff before private stuff.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  // Type of records to write
+  uint8_t Type = SRecord::S1;
+  std::vector<SRecord> Records;
+  void writeSection(const SectionBase &S, ArrayRef<uint8_t> Data);
```
Please split function members from object members by a blank line.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  std::vector<SRecord> Records;
+  void writeSection(const SectionBase &S, ArrayRef<uint8_t> Data);
+
+public:
```
As above: move public before protected/private stuff.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  // Once the type of all records is known, write the records
+  virtual void writeRecords(uint32_t Entry);
+  uint64_t getBufferOffset() const { return Offset; }
+  Error visit(const Section &S) final;
```
It's incredibly unusual to use `final` in my experience, and rarely, if ever, adds anything, so I'd get rid of it. Just use `override` as appropriate.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+  Error visit(const DynamicRelocationSection &S) final;
+  uint8_t getType() const { return Type; };
+  void writeRecord(SRecord &Record);
+  using BinarySectionWriter::visit;
```
I'd put this before the functions, separated by a blank line to make it stand out more.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -390,6 +392,104 @@ class IHexWriter : public Writer {
+/// Base class for SRecSectionWriter
+/// This class does not actually write anything. It is only used for size
+/// calculation.
+class SRECSectionWriterBase : public BinarySectionWriter {
```
Is a separate base class really needed? There is only one subclass as far as I can tell.

quic-areg replied:
I think it is useful to have the base class be used for size calculation and have the concrete class do the real writing. The two have different implementations of `writeRecords` and the correct function is called appropriately.

jh7370 replied:
This doesn't look or sound like it's describing an "is a" relationship to me, which is what inheritance should be used for. It sounds more like you've got two separate classes which share some common behaviour, which would mean you should have a common abstract base class that they both share, with two concrete sub-classes (or two independent classes that use a third class, but don't inherit from it).

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/Inputs/srec-elf-sections-err.yaml:
```diff
@@ -0,0 +1,22 @@
+!ELF
```
This file is only used in one place. Please inline it into that test, so that the test input data can be found with the rest of the test.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/Inputs/srec-elf-sections-err.yaml:
```diff
@@ -0,0 +1,22 @@
+    Type:            SHT_PROGBITS
+    Flags:           [ SHF_ALLOC, SHF_EXECINSTR ]
+    Address:         0xFFFFFFF8
+    AddressAlign:    0x8
```
`AddressAlign` is unnecessary for the test, I believe, so please delete it.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/Inputs/srec-elf-sections-err.yaml:
```diff
@@ -0,0 +1,22 @@
+# Part of section data is in 32-bit address range
+# and part isn't.
+    Type:            SHT_PROGBITS
+    Flags:           [ SHF_ALLOC, SHF_EXECINSTR ]
```
`SHF_EXECINSTR` is unnecessary for the test, I believe, so please delete it (same below).

Keep your test data to a minimum for what your test actually needs.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/Inputs/srec-elf-sections.yaml:
```diff
@@ -0,0 +1,53 @@
+  Machine:         EM_X86_64
+Sections:
+  - Name:            .data1
+# Records for this section should come last
```
Nit: comment style applies to tests too.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/Inputs/srec-elf-sections.yaml:
```diff
@@ -0,0 +1,53 @@
+!ELF
```
To inline this with other YAML files in the same test file, I think you'll need to prefix this with `---`

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/Inputs/srec-elf-segments.yaml:
```diff
@@ -0,0 +1,60 @@
+# Test copied from ihex-elf-segments.yaml:
```
The relationship with the ihex test is largely irrelevant. I shouldn't need to look at a different test to understand the logic of why this input contains what it does. Instead, properly comment on what is significant about this input (and delete any parts of the input that aren't relevant to the test).

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -0,0 +1,70 @@
+# RUN: yaml2obj %p/Inputs/ihex-elf-segments.yaml -o %t.seg
+# RUN: llvm-objcopy -O srec %t.seg - | FileCheck %s --check-prefix=PADDR
+
+## # The record type for the header should be S0 with a 2 byte address
```
What's going on with the start of the comment here?

jh7370 reviewed (changes requested):
Overall, this doesn't look too bad, but there's a lot of tidying up that needs to happen before I'd be happy for this to be merged in.

quic-areg commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2813,6 +2813,244 @@ Error IHexWriter::finalize() {
+Error SRECSectionWriterBase::visit(const StringTableSection &Sec) {
+  // Check that sizer has already done its work
+  assert(Sec.Size == Sec.StrTabBuilder.getSize());
+  // we don't need to write anything here the real writer has already done it.
```
Done

quic-areg commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2813,6 +2813,244 @@ Error IHexWriter::finalize() {
+
+void SRECSectionWriterBase::writeRecords(uint32_t Entry) {
+  // The ELF header could contain an entry point outside of the sections we have
+  // seen that does not fit the current record Type
```
Done

quic-areg commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2813,6 +2813,244 @@ Error IHexWriter::finalize() {
+  return Line;
+}
+
+uint8_t SRecord::getChecksum() const {
```
The best source I found is the wikipedia page and its citations: https://en.wikipedia.org/wiki/SREC_(file_format)#Checksum_calculation

jh7370 reviewed (commented):
@quic-areg, please don't resolve comment threads where I've requested changes: please let me resolve them as a marker that I'm happy with the resolution.

A thought on testing: how do you show that the lines are terminated properly (including a `\r`)?

quic-areg commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -399,95 +423,81 @@ struct SRecord {
-  // Once the type of all records is known, write the records
+  using BinarySectionWriter::visit;
+
+  // Once the type of all records is known, write the records.
```
Removed

quic-areg commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -399,95 +423,81 @@ struct SRecord {
+  using BinarySectionWriter::visit;
+
+  // Once the type of all records is known, write the records.
   virtual void writeRecords(uint32_t Entry);
```
No it does not override anything. It is now a pure virtual function as part of the refactoring you suggested.

quic-areg replied:
Removed

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -45,26 +29,164 @@
+  Machine:         EM_X86_64
+Sections:
+  - Name:            .data1
+# Records for this section should come last.
```
Done

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -1,18 +1,6 @@
-# RUN: llvm-objcopy -O srec --only-section=.text %t.err - \
-#   RUN:| FileCheck %s --check-prefix=ZERO_SIZE_SEC
-
 ## Terminator should contain the entry point
```
Done

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -29,11 +17,7 @@
-
-## # The record type for the header should be S0 with a 2 byte address
+## The record type for the header should be S0 with a 2 byte address
 ## of 0. For an output file named "-" the header data field should contain "2D"
```
Done

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -45,26 +29,164 @@
+# BAD-ADDR: section '.text1' address range [0xfffffff8, 0x100000000] is not 32 bit
+# BAD-ADDR2: section '.text2' address range [0xffffffff0, 0xffffffff4] is not 32 bit
 
 # there should be no records besides header and terminator
```
Done

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -45,26 +29,164 @@
+# BAD-ADDR2: section '.text2' address range [0xffffffff0, 0xffffffff4] is not 32 bit
 
 # there should be no records besides header and terminator
 # ZERO_SIZE_SEC-NOT: {{S[1-8]}}
```
Used the latter approach

quic-areg commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -45,26 +29,164 @@
+## triggers an error
+# RUN: yaml2obj %s --docnum=2 -o %t.err
+# RUN: not llvm-objcopy -O srec --only-section=.text1 %t.err - 2>&1 \
+#   RUN: | FileCheck %s --check-prefix=BAD-ADDR
```
Done

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2987,7 +2987,7 @@ size_t SRECWriter::writeTerminator(uint8_t *Buf, uint8_t Type) {
 Expected<size_t>
 SRECWriter::getTotalSize(WritableMemoryBuffer &EmptyBuffer) const {
-  SRECSectionWriterBase LengthCalc(EmptyBuffer, 0);
+  SRECSizeCalculator LengthCalc(EmptyBuffer, 0);
```
Nit: it feels a bit off to be mixing the terms "Length" and "Size" for the same thing. I don't mind which you use (although I think "Size" is more common in general), but I'd be consistent.

quic-areg replied:
Done

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2847,24 +2847,24 @@ void SRECSectionWriter::writeRecord(SRecord &Record, uint64_t Off) {
 }
 
-void SRECSectionWriter::writeRecords(uint32_t Entry) {
+void SRECSizeCalculator::writeRecords(uint32_t Entry) {
```
Taking another look at this and it's not really ideal that we have to do two passes through the records (one for sizing and one for writing). I'm sure there's a reason to do it this way, but what is it, as that might inform an alternative approach?

quic-areg replied:
We can write to the output buffer without resizing or copying.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2847,24 +2847,24 @@ void SRECSectionWriter::writeRecord(SRecord &Record, uint64_t Off) {
-  // The ELF header could contain an entry point outside of the sections we have
-  // seen that does not fit the current record Type.
+void SRECSectionWriter::writeRecords(uint32_t Entry) {
   Type = std::max(Type, SRecord::getType(Entry));
```
Could we move this function into the base class, and just make `writeRecord` the sub-class specific method (in the size calculation case, it does nothing)?

quic-areg replied:
Done

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -1,36 +1,40 @@
+## Check for basic functionality using an input file with
+## various section types, adresses, data, and no segments.
+
```
Nit: as the comment above is tightly-linked to the RUN lines immediately below, I'd get rid of the blank line.

quic-areg replied:
Done

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -1,36 +1,40 @@
-# ENTRY: S00400002DCE
-# ENTRY-NEXT: S705F00000000A
+## 1 byte for output file and 1 byte for checksum.
+     # SREC:S00400002DCE
```
Let's move the CHECK patterns to be close to the individual cases they are relevant to, so your ordering goes RUN/CHECK/RUN/CHECK etc.

quic-areg replied:
Done

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -44,64 +48,65 @@ FileHeader:
-# This section's contents exceeds default line length of 16 bytes
-# so we expect two lines created for it. Records for this section
-# should appear before records for the previous section.
+## This section's contents exceeds default line length of 16 bytes
```
Nit
```suggestion
## This section's contents exceed default line length of 16 bytes
```
Nit: "content exceeds" or "contents exceed" to be grammatically correct.

quic-areg replied:
Done

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -111,36 +116,39 @@ FileHeader:
   - Name:            .text1
-# Part of section data is in 32-bit address range
-# and part isn't.
+## Part of section data is in 32-bit address range
```
Nit: Unnecessary line-wrapping (use 80-character width).

quic-areg replied:
Done

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/srec-writer.test:
```diff
@@ -111,36 +116,39 @@ FileHeader:
-# RUN: llvm-objcopy -O srec %t.seg - | FileCheck %s --check-prefix=PADDR
+## This tests an input file with segments and expects
+## physical addresses instead of virtual addresses.
 
```
Nit: as above, delete this blank line to tie the comment to the associated RUN lines.

quic-areg replied:
Done

jh7370 reviewed (approved):
Aside from one nit (unnecessary `virtual` I believe), I think this PR now gets an LGTM from me.

A couple of caveats:

- I haven't attempted to review this against the SREC docs: I don't have time to do this, so I'm assuming that @statham-arm has done so.
- I haven't attempted to verify that the testing covers everything. If you have access to a code coverage tool, it might be instructive to run it through your modified llvm-objcopy with the tests you've written, to see if you've covered all the new code you've added.

P.S. Sorry for the delay - a combination of illness and a large review backlog has delayed me getting back to this until now.

quic-areg commented:
Can this be merged? I have squashed my commits but I do not think I have permissions to merge.

jh7370 reviewed (approved):
LGTM. I'll look to merge shortly.

P.S. there's no need to manually squash this as we use the Squash & Merge approach in this project (i.e. squashing happens automatically).

jh7370 commented:
@quic-areg, it looks like you have your email address hidden through github. Per the consensus on https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223/32 this isn't really desired. You have two options:
1) You change github to not hide your email address.
2) Send me your name and email address so that I can push this directly.
Please let me know if you do 1, so that someone can then merge this PR at that point.

quic-areg commented:
@jh7370 Github should show my email now

jh7370 commented:
Thanks, merged. Please keep an eye out for emails from the build bots: if you get any failure notifications, check to see if it is likely to do with your change.

jroelofs commented:
build fix for this in 5afbed1968588fe443a8a55053c2f1eaa321d28e

maksfb commented:
Another fix in 2cbe5a33a5fda257747d75863bd9ccb8920b9249
